org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/MethodBinding.java

```diff
@@ -73,6 +73,8 @@ public class MethodBinding extends Binding {
 	public Boolean[] parameterNonNullness;  // TRUE means @NonNull declared, FALSE means @Nullable declared, null means nothing declared
 	public int defaultNullness; // for null *type* annotations
 
+	public Boolean[] parameterOwning; // TRUE means @Owning declared, ...
```

Comment:
I can understand why `parameterNonNullness` benefits from the boxed Boolean. For `parameterOwning` I don't see why we cannot use `boolean[]`

Does it make sense to bit-encode the information of both arrays into a single long[]?

Reply:
I haven't yet dismissed the idea to add the dual annotation `@NotOwning`, in particular because having that additional option would allow us to make `@Owning` the default for method returns.

But while consolidating the implementation, merging all this into a array of bitsets makes perfect sense. Assuming that you didn't only consider beauty of the code, but at least partially performance, too, I went one step further to choose `byte[]` for now. This means we cannot directly use `TagBits.AnnotationOwning` and friends as index, but we are converting between various formats anyway, and conversions look pretty succinct to me.